rest_api/controller/search.py

```diff
@@ -65,7 +67,7 @@ def query(request: QueryRequest):
         return result
 
 
-def _process_request(pipeline, request) -> QueryResponse:
+def _process_request(pipeline, request) -> Dict[str, Any]:
```

Comment:
Why can't we keep `QueryResponse` here? Line 85 looks like we are making sure it's a `QueryResponse`.

Reply:
Yes I thought the same. However, in practice we were returning a dictionary all the time and worse yet, returning a real `QueryResponse` causes a Pydantic validation error. So I decided to change the type to clarify this.

I think the type on line 85 is needed by Pydantic to figure out which object to use for validation, but I'm not completely sure.

Reply:
Let's try to improve on this. The `Dict[str, Any]` is  also inconsistent with the response model in  `@router.post("/query", response_model=QueryResponse, response_model_exclude_none=True)`.
@tholor do you have some advice how to make use of `QueryResponse`here?

Reply:
I had a look at the FastAPI docs and it looks like it's designed to handle with dictionaries without any issue: https://fastapi.tiangolo.com/tutorial/response-model/ However I agree it's odd that it was failing validation when returning a `QueryResponse`.

Reply:
The most important thing is that we declare `QueryResponse` as the response model on the real endpoints (enables openapi/swagger documentation, validation ...). 

On this "helper method" `_process_request`: I agree that the old version didn't make sense (returning a dict, but annotating with QueryResponse). I think it would make most sense to change the actual returned object to a QueryResponse - What pydantic error did you get there when trying it? 

If this is a rabbit hole, switching the return type annotation to Dict is also not too bad here IMO as we still enforce the response model on the higher-level endpoint.

I don't see much value in adding this "conversion-and-back-to-dict step" in line 85. This will happen anyway on-the-fly in the endpoint and is just extra computation time :thinking:

Reply:
Pydantic errors are always hard to figure out. If I implement the method this way:
```python
def _process_request(pipeline, request) -> QueryResponse:
    ...
    result = pipeline.run(query=request.query, params=params, debug=request.debug)
    response = QueryResponse(**result)
    ...
    return response
```
I get this error on every tests that calls `/query`:

![2022-02-11_14:28:44](https://user-images.githubusercontent.com/5703634/153600039-5cb08db4-80f2-4cc6-9a37-176562c3d7d9.png)

I pushed a commit with the method implemented this way so you can see yourself how the error looks like. For me is really hard to parse :sweat_smile:

Reply:
Here the actual errors: https://github.com/deepset-ai/haystack/runs/5156855217?check_suite_focus=true

Reply:
I went back to the dictionary type because I still can't find a way to make `QueryResponse` work, and I think it's a minor issue anyway. Let's keep it like this for now.